### ISSUE TITLE
MakeFieldReadonlyDiagnosticAnalyzer does not report diagnostic when used via CompilationWithAnalyzers

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#41523]
**Version Used**: 3.4.0 nupgk

The only way a console app can run Roslyn analyzers is through the CompilationWithAnalyzers methods. When `MakeFieldReadonlyDiagnosticAnalyzer` is referenced and run this way, it does not report IDE0044 when it should. When Visual Studio runs `MakeFieldReadonlyDiagnosticAnalyzer` for the same solution and file, IDE0044 shows up in the error list.

A central need of https://github.com/jnm2/SuppressionCleanupTool is the ability to tell which IDExxxx diagnostics Visual Studio will show in the error list.

Self-contained repro:

#### Program.cs

<details>

```cs
using Microsoft.Build.Locator;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.MSBuild;
using System;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading;
using System.Threading.Tasks;

public static class Program
{
    class ClassTriggeringIDE0044
    {
        // VS shows 'IDE0044 Make field readonly Roslyn' here
        //          ↓
        private int foo;

        public ClassTriggeringIDE0044(int foo) => this.foo = foo;
    }

    public static async Task Main()
    {
        MSBuildLocator.RegisterDefaults();

        using (var workspace = MSBuildWorkspace.Create())
        {
            var solution = await workspace.OpenSolutionAsync(@"C:\Users\Joseph\Source\Repos\RoslynRepro\RoslynRepro.sln");
            var project = solution.Projects.Single();
            var document = project.Documents.Single(d => Path.GetFileName(d.FilePath) == "Program.cs");

            var csharpFeaturesAssembly = Assembly.Load("Microsoft.CodeAnalysis.Features");
            var csharpFeaturesAnalyzerReference = new AnalyzerFileReference(
                csharpFeaturesAssembly.Location,
                new SingleAnalyzerAssemblyLoader(csharpFeaturesAssembly));

            var analyzers = csharpFeaturesAnalyzerReference.GetAnalyzers(project.Language)
                .Where(instance => instance.GetType().Name == "MakeFieldReadonlyDiagnosticAnalyzer")
                .ToImmutableArray();

            var semanticModel = await document.GetSemanticModelAsync();

            var compilationWithAnalyzers = semanticModel.Compilation.WithAnalyzers(analyzers, new CompilationWithAnalyzersOptions(
                project.AnalyzerOptions,
                onAnalyzerException: null,
                concurrentAnalysis: false,
                logAnalyzerExecutionTime: true,
                reportSuppressedDiagnostics: true));

            var diagnostics = await compilationWithAnalyzers.GetAnalyzerSemanticDiagnosticsAsync(semanticModel, filterSpan: null, CancellationToken.None);

            Console.WriteLine(diagnostics.Length); // Prints 0
        }
    }

    private sealed class SingleAnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
    {
        private readonly Assembly assembly;

        public SingleAnalyzerAssemblyLoader(Assembly assembly)
        {
            this.assembly = assembly;
        }

        public void AddDependencyLocation(string fullPath)
        {
        }

        public Assembly LoadFromPath(string fullPath)
        {
            if (!string.Equals(fullPath, assembly.Location, StringComparison.OrdinalIgnoreCase))
                throw new InvalidOperationException($"This assembly loader must only load '{assembly.Location}'.");

            return assembly;
        }
    }
}
```

</details>

##### Repro.csproj

<details>

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net48</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Build.Locator" Version="1.2.6" />
    <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.MSBuild" Version="3.4.0" />
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Features" Version="3.4.0" />
  </ItemGroup>

</Project>
```

</details>

**Comment by jnm2** [comment]
Looks like the same thing is happening for these analyzers as well:

- IDE0062 Local function can be made static (MakeLocalFunctionStaticDiagnosticAnalyzer)
- IDE0059 Unnecessary assignment of a value
  (CSharpRemoveUnusedParametersAndValuesDiagnosticAnalyzer)
- IDE0031 Null check can be simplified
  (CSharpUseNullPropagationDiagnosticAnalyzer)

**Comment by jnm2** [comment]
@CyrusNajmabadi pointed out that this might be related to https://github.com/dotnet/roslyn/issues/41521 in that `GetDocumentOptionSetAsync` is returning null. I didn't debug to verify this, but it seems likely that is what is happening in all four cases in this issue as well:

<details>

https://github.com/dotnet/roslyn/blob/5858196e0196bcfaf9887e28efdfd500a9fda122/src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs#L145-L148

https://github.com/dotnet/roslyn/blob/5858196e0196bcfaf9887e28efdfd500a9fda122/src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs#L87-L97

https://github.com/dotnet/roslyn/blob/5858196e0196bcfaf9887e28efdfd500a9fda122/src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs#L83-L90

https://github.com/dotnet/roslyn/blob/5858196e0196bcfaf9887e28efdfd500a9fda122/src/Features/CSharp/Portable/MakeLocalFunctionStatic/MakeLocalFunctionStaticDiagnosticAnalyzer.cs#L46-L56

</details>

If that is true, the question becomes: how do you use the public API (MSBuildWorkspace/OpenSolutionAsync/CompilationWithAnalyzers/GetAnalyzerSemanticDiagnosticsAsync) to run all analyzers with the default set of options?

**Comment by sharwell** [comment]
All of the analyzers in the features layer require `WorkspaceAnalyzerOptions`:

https://github.com/dotnet/roslyn/blob/5858196e0196bcfaf9887e28efdfd500a9fda122/src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs#L378

**Comment by jnm2** [comment]
I see that https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs is currently internal. Would you consider making a simple API in order to hook this all up?

**Comment by jnm2** [comment]
Until this is handled automatically or a new API is added, this workaround works for 3.4.0. Usage:

```cs
var analyzerOptions = TryCreateWorkspaceAnalyzerOptions(document.Project.AnalyzerOptions, document.Project.Solution)
    ?? throw new NotSupportedException("Cannot expose document options to analyzers to allow them to run as normal.");

var compilationWithAnalyzers = semanticModel.Compilation.WithAnalyzers(analyzers, analyzerOptions, cancellationToken);
```

Implementation:

```cs

public static AnalyzerOptions TryCreateWorkspaceAnalyzerOptions(AnalyzerOptions options, Solution solution)
{
    if (Type.GetType("Microsoft.CodeAnalysis.Diagnostics.WorkspaceAnalyzerOptions, Microsoft.CodeAnalysis.Features") is { } workspaceAnalyzerOptionsType)
    {
        const BindingFlags allInstance = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;

        // Roslyn 3.5 (changes: https://github.com/dotnet/roslyn/pull/40289)
        var constructor = workspaceAnalyzerOptionsType.GetConstructor(allInstance, null, new[] { typeof(AnalyzerOptions), typeof(Solution) }, null);
        if (constructor is { })
        {
            return (AnalyzerOptions)constructor.Invoke(new object[] { options, solution });
        }

        // Roslyn 2.3.2–3.2 (changes: https://github.com/dotnet/roslyn/pull/19277)
        constructor = workspaceAnalyzerOptionsType.GetConstructor(allInstance, null, new[] { typeof(AnalyzerOptions), typeof(OptionSet), typeof(Solution) }, null);
        if (constructor is { })
        {
            // Solution.Options is what Microsoft.CodeAnalysis.Diagnostics.AnalyzerHelper passed.
            return (AnalyzerOptions)constructor.Invoke(new object[] { options, solution.Options, solution });
        }

        // Roslyn 1.0–2.3.1
        constructor = workspaceAnalyzerOptionsType.GetConstructor(allInstance, null, new[] { typeof(AnalyzerOptions), typeof(Workspace) }, null);
        if (constructor is { })
        {
            // Solution.Workspace is what Microsoft.CodeAnalysis.Diagnostics.AnalyzerHelper passed.
            return (AnalyzerOptions)constructor.Invoke(new object[] { options, solution.Workspace });
        }
    }

    return null;
}
```

**Comment by sharwell** [comment]
> Would you consider making a simple API in order to hook this all up, perhaps making Project.AnalyzerOptions return an instance by default that implements GetDocumentOptionSetAsync properly?

Probably not right now. The goal is to move the IDE analyzers down to the code style layer, which would remove the workspace layer dependencies at the same time.
